### PR TITLE
Add strategy that optimize mixed index lookup followed by count step

### LINKED ISSUE #874: Add strategy that will optimize mixed index lookup followed by a count
I think we can make use of the work accomplished in #318 to optimize queries with the following structure `g.V().has("text", textFuzzy("foo")).count()`. The strategy should be able to replace the default counting behavior with one that uses the count retrieved from the index provider.

### PR DESCRIPTION
**Pull request by li-boxuan**
This adds JanusGraphMixedIndexCountStrategy which leverages mixed index
to return the total count if applicable. The implementation idea is to
leverage the partial result of JanusGraphStepStrategy and JanusGraphStep
which handle folding and building a global graph centric query.

Closes #874

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[doc only]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/IndexQueryBuilder.java [hunk 1]
```diff
@@ -152,6 +184,73 @@ public String getUnknownKeyName() {
...
+        if (Contain.IN.equals(constraint.getPredicate())) {
+            queryBuilder.append("(");
+            for (Object value : (List) constraint.getValue()) {
+                queryBuilder.append("\"" + value + "\"");
+            }
+            queryBuilder.append(")");
+        } else if (Text.CONTAINS.equals(constraint.getPredicate()) || Cmp.EQUAL.equals(constraint.getPredicate())) {
+            queryBuilder.append("\"" + constraint.getValue() + "\"");
+        } else {
+            // TODO: we should support all types as long as a single mixed query can be constructed
```
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/IndexQueryBuilder.java:249]: Now I think constructing a direct index query might be the wrong direction. Using existing `Stream<String> query(IndexQuery query, KeyInformation.IndexRetriever information, BaseTransaction tx)` in IndexProvider and appending `count` seems better.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMixedIndexCountStep.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    public MixedIndexCountQuery buildGlobalGraphIndexQuery(final GraphStep originalGraphStep, final JanusGraphTransaction tx) {
+        optProfiler = new TP3ProfileWrapper(new MutableMetrics(this.getId() + "optimization", "optimization"));
+        MixedIndexCountQueryBuilder countQueryBuilder = (MixedIndexCountQueryBuilder) tx.mixedIndexCountQuery();
+        List<PredicateCondition<String, JanusGraphElement>> constraints = new ArrayList<>();
+        for (final HasContainer condition : hasContainers) {
+            constraints.add(new PredicateCondition<>(condition.getKey(), JanusGraphPredicateUtils.convert(condition.getBiPredicate()), condition.getValue()));
+        }
+        return countQueryBuilder.constructIndex(optProfiler, constraints,
```
Comment (rngcntr) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMixedIndexCountStep.java:82]: I like the Idea that the step itself performs the time consuming task of index selection, which makes it possible to track the work done in the optimization phase. Great idea!

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMixedIndexCountStrategy.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            return;
+        }
+        final JanusGraphMixedIndexCountStep<?> directQueryCountStep = new JanusGraphMixedIndexCountStep<>(traversal);
+        // we shouldn't really fold in, but only check it, because we don't want to break the original traversal,
+        // as this optimizer might fail
+        if (!foldInHasContainer(originalGraphStep, directQueryCountStep, traversal, traversal)) {
+            return;
+        }
+        final JanusGraphTransaction tx = JanusGraphTraversalUtil.getTx(traversal);
+        MixedIndexCountQuery mixedIndexCountQuery = directQueryCountStep.buildGlobalGraphIndexQuery(originalGraphStep, tx);
```
Comment (rngcntr) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMixedIndexCountStrategy.java:79]: It took me about half an hour to figure out that _this_ is actually the line that checks if the optimization should be applied at all. I tried to find the line that clearly differentiates between mixed and composite indexes. The fact that this is buried deeply inside the index selection part is a bit of an obfuscation of what actually happens here.
Now that it became clear to me, it would be good to at least point out exactly what happens here.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMixedIndexCountStrategy.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        if (mixedIndexCountQuery != null) {
+            directQueryCountStep.setMixedIndexCountQuery(mixedIndexCountQuery);
+            removeHasContainerSteps(originalGraphStep, directQueryCountStep, traversal);
+        }
+    }
+
+    private boolean hasCountGlobalStep(final GraphStep originalGraphStep) {
+        Step<?, ?> currentStep = originalGraphStep.getNextStep();
+        while (currentStep instanceof IdentityStep || currentStep instanceof NoOpBarrierStep || currentStep instanceof HasContainerHolder) {
```
Comment (rngcntr) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMixedIndexCountStrategy.java:88]: These `instanceof` chains are used quite often so it would be better to move the check to a separate method. This would also give these steps a name that is more readable. For example `while (isEligibleToSkip(currentStep)) {...}`

**Review by rngcntr — COMMENTED** [review]
I took a look at your PR and really can't say much about it. It touches many components that I was unaware of before so I can't tell to which extent the introduced changes are appropriate. Left some comments though

**Comment by rngcntr** [comment]
One more thing: Have you measured the performance impact as you did for your previous PRs?

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java [hunk 5]
```diff
@@ -1112,6 +1113,99 @@ public void testIndexParameters() {
...
+    public void testMixedIndexQueryFollowedByCount() {
+        final int numV = 1000;
+        long total;
+        TraversalMetrics metrics;
+        final String[] strings = {"Uncle Berry has a farm", "and on his farm he has five ducks", "ducks are beautiful animals", "the sky is very blue today"};
+        setupChainGraph(numV, strings, true);
+        clopen();
+
+        /* ===================================================
+        has(key, value) followed by count() can be optimized */
```
[the remark this answers is not in the document]
Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1130]: Totally. I need to add more test cases.

**Comment by porunov** [comment]
@li-boxuan Do you have plans to work on this PR? Would be great to see this feature as many users are not aware of direct index queries

**Comment by li-boxuan** [comment]
@porunov Thanks for reminder! I already forgot about this PR 😅 I will try to find some time to work on it this weekend, but please don’t count it as a blocker for 0.6.0 release.

**Comment by li-boxuan** [comment]
Based on https://github.com/JanusGraph/janusgraph/pull/2684/

**Comment by porunov** [comment]
@li-boxuan Looks like "Reduce redundant global centric query build" commit is still in this PR

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java [hunk 6]
```diff
@@ -1523,6 +1525,203 @@ public void testIndexParameters() {
...
+
+        // vertex
+        total = numV / strings.length * 2;
+        assertEquals(total, graph.indexQuery(VINDEX, "v.text:ducks").vertexTotals());
+        TraversalMetrics profile = graph.traversal().V().has("text", Text.textContains("ducks")).count().profile().next();
+        final String xtextKey = "xtext";
```
Comment (porunov) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1557]: To pass Solr tests it probably should be:
```
final String xtextKey = getStringField("xtext");
```

Maybe in some other places as well. I didn't check the whole test yet

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        TraversalHelper.replaceStep(originalGraphStep, mixedIndexCountStep, traversal);
+    }
+
+    public static JanusGraphMixedIndexCountStrategy instance() {
+        return INSTANCE;
+    }
+
+    @Override
+    public Set<Class<? extends ProviderOptimizationStrategy>> applyPrior() {
+        return Collections.singleton(JanusGraphStepStrategy.class);
```
Comment (porunov) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java:119]: Looks like this one can actually be a constant (i.e. `private final static`)

**Review by porunov — APPROVED** [review]
LGTM. Thank you @li-boxuan for such a great improvement! No need to use direct index queries for count queries anymore :)
I have only one very small nitpick which not necessary to be resolved.

**Comment by li-boxuan** [comment]
@rngcntr Do you want to review again?

**Comment by li-boxuan** [comment]
Using the following code to do a simple performance test (single Cassandra + single Elasticsearch)

```java
public class JanusGraphCqlApp {
    public static void main(String[] args) {
        Map<String, String> map = new HashMap() {{
            put("storage.backend", "cql");
            put("storage.hostname", "127.0.0.1");
            put("storage.cql.keyspace", "janusgraph");
            put("index.search.backend", "elasticsearch");
            put("index.search.hostname", "127.0.0.1");
        }};
        MapConfiguration config = new MapConfiguration(map);
        config.setListDelimiterHandler(new DefaultListDelimiterHandler(','));
        JanusGraph graph = JanusGraphFactory.open(config);
        JanusGraphManagement mgmt = graph.openManagement();
        PropertyKey stringKey = mgmt.makePropertyKey("name").dataType(String.class).make();
        mgmt.buildIndex("namev", Vertex.class).addKey(stringKey, Mapping.STRING.asParameter())
            .buildMixedIndex("search");
        mgmt.commit();
        for (int i = 0; i < 100; i++) {
            for (int j = 0; j < 10000; j++) {
                graph.addVertex("name", "value");
            }
            graph.tx().commit();
        }

        // query
        long startTime = System.currentTimeMillis();
        long count = graph.traversal().V().has("name", "value").count().next();
        System.out.println("count = " + count + ", time spent (ms) = " + (System.currentTimeMillis() - startTime));
        graph.close();
    }
}
```

master:

count = 1000000, time spent (ms) = 318092
count = 1000000, time spent (ms) = 399118
count = 1000000, time spent (ms) = 315376


with this PR:

count = 1000000, time spent (ms) = 181
count = 1000000, time spent (ms) = 210
count = 1000000, time spent (ms) = 172

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java [hunk 8]
```diff
@@ -1523,6 +1529,202 @@ public void testIndexParameters() {
...
+
+        profile = graph.traversal().V().has("age", 0).count().profile().next();
+        assertFalse(profile.getMetrics(0).getName().contains(JanusGraphMixedIndexCountStep.class.getSimpleName()));
+        assertTrue(profile.getMetrics(0).getName().contains(JanusGraphStep.class.getSimpleName()));
+        assertEquals(5, graph.traversal().V().has("age", 0).count().next());
+        profile = graph.traversal().V().has("age").count().profile().next();
+        assertFalse(profile.getMetrics(0).getName().contains(JanusGraphMixedIndexCountStep.class.getSimpleName()));
+        assertTrue(profile.getMetrics(0).getName().contains(JanusGraphStep.class.getSimpleName()));
+        assertEquals(10, graph.traversal().V().has("age").count().next());
+
```
Comment (porunov) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1725]: I was thinking about more test cases. Can we add 2 more test cases?
`has("nonIndexedProperty").has("indexedProperty").count()`
`has("indexedProperty").has("nonIndexedProperty").count()`
What do you think about it? Maybe we already checking these cases in other tests. If so, just ignore this comment 
the above 2 queries are computed the same but they should not be optimized with Mixed Query Count optimization because the whole chain of 'has' steps cannot be answered by a single mixed index.

**Review by porunov — APPROVED** [review]

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java [hunk 9]
```diff
@@ -1124,6 +1126,7 @@ public void testGraphCentricQueryProfiling() {
         if (indexFeatures.supportNotQueryNormalForm()) {
             newTx();
             assertEquals(3, tx.traversal().V().or(__.has("name", "bob"), __.has("prop", "val")).count().next());
+            assertEquals(3, tx.traversal().V().or(__.has("name", "bob"), __.has("prop", "val")).toList().size());
```
Comment (rngcntr) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1129]: What is the reason to change these queries from `count().next()` to `toList().size()`? I see that the result should be equivalent, but therefore I don't see the necessity to change the query.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+     *
+     * @param originalGraphStep
+     * @param mixedIndexCountStep
+     * @param traversal
+     */
+    private void applyMixedIndexCountStep(final GraphStep originalGraphStep, final JanusGraphMixedIndexCountStep mixedIndexCountStep,
+                                          final Traversal.Admin<?, ?> traversal) {
+        Step<?, ?> currentStep = originalGraphStep.getNextStep();
+        while (true) {
+            if (currentStep instanceof HasContainerHolder || currentStep instanceof CountGlobalStep) {
```
Comment (rngcntr) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java:104]: I stumbled upon this logic and was wondering if it is correct. For example, a query with `count().count()` (which by the way is completely bonkers) should always return 1, but would be reduced to one `JanusGraphMixedIndexCountStep` here.

**Review by rngcntr — COMMENTED** [review]
Hi @li-boxuan! I reviewed your changes again and I think I would feel better releasing this if there were also tests combining mixed and composite indexes. Other than that, there are two questions that came to my mind:

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java [hunk 11]
```diff
@@ -1523,6 +1529,213 @@ public void testIndexParameters() {
...
+        assertTrue(profile.getMetrics(0).getName().contains(JanusGraphStep.class.getSimpleName()));
+        assertEquals(10, graph.traversal().V().has("age").count().next());
+
+        /* =============================================================================
+        count() cannot be optimized if a single mixed index cannot meet all conditions */
+
+        assertEquals(10, graph.traversal().V().has("name").has("age").count().next());
+        profile = graph.traversal().V().has("name").has("age").count().profile().next();
+        assertFalse(profile.getMetrics(0).getName().contains(JanusGraphMixedIndexCountStep.class.getSimpleName()));
+        assertEquals(5, graph.traversal().V().has("name").has("age", 0).count().next());
```
Comment (li-boxuan) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1732]: @rngcntr This covers the case "tests combining mixed and composite indexes". "age" is indexed by a composite index while "name" is indexed by a mixed index.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java (hunk 9 above)
Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1129]: Now that count() queries are optimized by `JanusGraphMixedIndexCountStrategy`, I just add this extra check to ensure correctness when JanusGraphMixedIndexCountStrategy does not apply.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java (hunk 10 above)
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java:104]: That is a great point! I don't even know `count().count()` is legal. I'll check if the result is still correct in this case.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java (hunk 11 above)
Comment (rngcntr) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1732]: Oh, I overlooked that one. Well in this case, just ignore my comment ;)

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java (hunk 10 above)
Comment (rngcntr) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java:104]: Should be legal because `count()` simply produces a single traverser with a `count` value. This traverser itself can also be counted even if that makes no sense.
In addition, I am not entirely sure if there can exist other `HasContainerHolder` steps between the `JanusGraphStep` and the `CountGlobalStep`. For example, `JanusGraphStep` itself is a `HasContainerHolder` and so may be other steps. Let me take something like this as an example: `g.V().V().count()`. Since both `V()` are represented by `JanusGraphStep`, both satisfy the condition for this strategy.

Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java]: I rewrote this method and added a few test cases. Let me know if you come up with other corner cases.

**Comment by porunov** [comment]
@li-boxuan there is a checkstyle error due to unused import
```
Error:  /home/runner/work/janusgraph/janusgraph/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexCountStrategy.java:20:8: Unused import - org.apache.tinkerpop.gremlin.process.traversal.step.HasContainerHolder. [UnusedImports]
```

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java [hunk 12]
```diff
@@ -1523,6 +1529,223 @@ public void testIndexParameters() {
...
+        profile = graph.traversal().V().has("name").has("age", 0).count().profile().next();
+        assertFalse(profile.getMetrics(0).getName().contains(JanusGraphMixedIndexCountStep.class.getSimpleName()));
+        profile = graph.traversal().V().has("age").has("name").count().profile().next();
+        assertFalse(profile.getMetrics(0).getName().contains(JanusGraphMixedIndexCountStep.class.getSimpleName()));
+
+        /* ==========================
+        verify other special cases */
+
+        assertEquals(1, graph.traversal().V().has("name").count().count().next());
+        assertEquals(numV, graph.traversal().V().V().has("text").count().next());
```
Comment (rngcntr) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1742]: I don't think this would be the expected behavior for chained `V()` calls. Instead, `V()` should spawn a new traverser at every single vertex for each incoming traverser. Thus, for a graph with `100` vertices, `g.V().V().count()` would return `10.000`. With this test however, you seem to verify that the result is `100`.

I suggest to test the correctness by comparing the result of `g.[...]` to `g.withoutStrategies(JanusGraphMixedIndexCountStrategy.class).[...]`. That would make sure that under no circumstances the optimization affects the output result. In my opinion, this should even be a considered standard procedure when introducing new optimization strategies. (Not blaming you, I skipped this kind of verification for some of my optimization strategies, too)

Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1742]: Great catch! I'll fix this.

Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java]: Now if JanusGraphStep is not the start step, JanusGraphMixedIndexCountStrategy won't be applied.

Comment (rngcntr) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java]: That looks much better, thank you @li-boxuan

**Review by rngcntr — APPROVED** [review]
